compiler/qsc_qasm3/src/tests/statement/const_eval.rs

```diff
@@ -0,0 +1,1483 @@
+        mutable r = [Zero, Zero, Zero, Zero];
+    "#]]
+    .assert_eq(&qsharp);
+    Ok(())
+}
+
+// Shr
+
+#[test]
+fn binary_op_shr_uint() -> miette::Result<(), Vec<Report>> {
+    let source = r#"
+        const uint a = 5;
+        const uint b = a >> 2;
+        bit[b] r;
+    "#;
+
+    let qsharp = compile_qasm_to_qsharp(source)?;
+    expect![[r#"
+        let a = 5;
+        let b = a >>> 2;
+        mutable r = [Zero, Zero, Zero, Zero];
+    "#]]
+    .assert_eq(&qsharp);
+    Ok(())
+}
+
+#[test]
+#[ignore = "angles are not yet supported"]
+fn binary_op_shr_angle() -> miette::Result<(), Vec<Report>> {
+    let source = r#"
+        const angle a = 1;
+        const angle b = a >> 2;
+        const bool c = b;
+        bit[c] r;
+    "#;
+
+    let qsharp = compile_qasm_to_qsharp(source)?;
+    expect![[r#""#]].assert_eq(&qsharp);
+    Ok(())
+}
+
+#[test]
+fn binary_op_shr_bit() -> miette::Result<(), Vec<Report>> {
+    let source = r#"
+        const bit a = 1;
+        const bit b = a >> 2;
+        bit[b] r;
+    "#;
+
+    let qsharp = compile_qasm_to_qsharp(source)?;
+    expect![[r#"
+        let a = 1;
+        let b = a >>> 2;
+        mutable r = [];
+    "#]]
+    .assert_eq(&qsharp);
+    Ok(())
+}
+
+#[test]
+fn binary_op_shr_bitarray() -> miette::Result<(), Vec<Report>> {
+    let source = r#"
+        const bit[4] a = "1011";
+        const bit[4] b = a >> 2;
+        bit[b] r;
+    "#;
+
+    let qsharp = compile_qasm_to_qsharp(source)?;
+    expect![[r#"
+        let a = 1;
+        let b = a <<< 2;
+        mutable r = [Zero, Zero];
+    "#]]
+    .assert_eq(&qsharp);
+    Ok(())
+}
+
+// BinaryOp: Bitwise
+
+// AndB
+
+#[test]
+fn binary_op_andb_uint() -> miette::Result<(), Vec<Report>> {
+    let source = r#"
+        const uint a = 5;
+        const uint b = a & 6;
+        bit[b] r;
+    "#;
+
+    let qsharp = compile_qasm_to_qsharp(source)?;
+    expect![[r#"
+        let a = 5;
+        let b = a &&& 6;
+        mutable r = [Zero, Zero, Zero, Zero];
+    "#]]
+    .assert_eq(&qsharp);
+    Ok(())
+}
+
+#[test]
+#[ignore = "angles are not yet supported"]
+fn binary_op_andb_angle() -> miette::Result<(), Vec<Report>> {
+    let source = r#"
+        const angle a = 1;
+        const angle b = a & 2;
+        const bool c = b;
+        bit[c] r;
+    "#;
+
+    let qsharp = compile_qasm_to_qsharp(source)?;
+    expect![[r#""#]].assert_eq(&qsharp);
+    Ok(())
+}
+
+#[test]
+fn binary_op_andb_bit() -> miette::Result<(), Vec<Report>> {
+    let source = r#"
+        const bit a = 1;
+        const bit b = a & 0;
+        bit[b] r;
+    "#;
+
+    let qsharp = compile_qasm_to_qsharp(source)?;
+    expect![[r#"
+        let a = 1;
+        let b = a &&& 0;
+        mutable r = [];
+    "#]]
+    .assert_eq(&qsharp);
+    Ok(())
+}
+
+#[test]
+fn binary_op_andb_bitarray() -> miette::Result<(), Vec<Report>> {
+    let source = r#"
+        const bit[4] a = "1011";
+        const bit[4] b = a & "0110";
```

Comment:
Do we need corresponding tests for:
```rust
        const uint[4] a = 11;
        const bit[4] b = a & "0110";
```

Same follows for `|` and `^`. Do we also need (I think this is valid):
```rust
        const uint[4] a = 11;
        const bit b = a & "0110";
```